File: 1-adaptive-bitrate-streaming-hls-dash/0-typescript/src/stream/stream.controller.ts
import { Controller, Get, Param, Post, Res } from '@nestjs/common';
import { Response } from 'express';
import { StreamService } from './stream.service';

/**
 * HLS endpoints. Same contract in all four languages:
 *   POST /api/stream/:id/encode          -> 200 { id, manifest, variants }
 *   GET  /api/stream/:id/manifest.m3u8   -> 200 application/vnd.apple.mpegurl
 *   GET  /api/stream/:id/:quality/index.m3u8 -> 200 variant playlist
 *   GET  /api/stream/:id/:quality/:segment   -> 200 video/mp2t (404 if missing)
 */
@Controller('api/stream')
export class StreamController {
  constructor(private readonly streamService: StreamService) {}

  @Post(':id/encode')
  async encode(@Param('id') id: string, @Res() res: Response): Promise<void> {
    await this.streamService.ensureTranscoded(id);
    res.status(200).json({
      id,
      manifest: `/api/stream/${id}/manifest.m3u8`,
      variants: ['360p', '720p', '1080p'],
    });
  }

  @Get(':id/manifest.m3u8')
  async manifest(@Param('id') id: string, @Res() res: Response): Promise<void> {
    // Lazy transcode on first manifest request, then serve from disk.
    await this.streamService.ensureTranscoded(id);
    res.setHeader('Content-Type', 'application/vnd.apple.mpegurl');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.sendFile(this.streamService.masterPath(id));
  }

  @Get(':id/:quality/:segment')
  async segment(
    @Param('id') id: string,
    @Param('quality') quality: string,
    @Param('segment') segment: string,
    @Res() res: Response,
  ): Promise<void> {
    const filePath = this.streamService.resolveFile(id, quality, segment);
    if (!filePath) {
      res.status(404).json({ id, quality, segment, error: 'not found' });
      return;
    }
    res.setHeader('Access-Control-Allow-Origin', '*');
    if (segment.endsWith('.m3u8')) {
      res.setHeader('Content-Type', 'application/vnd.apple.mpegurl');
      res.setHeader('Cache-Control', 'no-cache');
    } else {
      // video/mp2t is the IANA media type for MPEG-2 Transport Stream segments.
      res.setHeader('Content-Type', 'video/mp2t');
      res.setHeader('Cache-Control', 'public, max-age=31536000, immutable');
    }
    res.sendFile(filePath);
  }
}

File: 1-adaptive-bitrate-streaming-hls-dash/1-java/src/main/java/com/starci/streaming/StreamController.java
package com.starci.streaming;

import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.http.CacheControl;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import java.util.Map;

/**
 * HLS endpoints. Same contract in all four languages:
 *   POST /api/stream/{id}/encode             -> 200 { id, manifest, variants }
 *   GET  /api/stream/{id}/manifest.m3u8      -> 200 application/vnd.apple.mpegurl
 *   GET  /api/stream/{id}/{quality}/index.m3u8 -> 200 variant playlist
 *   GET  /api/stream/{id}/{quality}/{seg}    -> 200 video/mp2t (404 if missing)
 */
@RestController
@RequestMapping("/api/stream")
public class StreamController {

    private final TranscodeService transcode;

    public StreamController(TranscodeService transcode) {
        this.transcode = transcode;
    }

    @PostMapping("/{id}/encode")
    public ResponseEntity<Map<String, Object>> encode(@PathVariable String id) throws Exception {
        transcode.ensureTranscoded(id);
        return ResponseEntity.ok(Map.of(
            "id", id,
            "manifest", "/api/stream/" + id + "/manifest.m3u8",
            "variants", List.of("360p", "720p", "1080p")
        ));
    }

    @GetMapping(value = "/{id}/manifest.m3u8", produces = "application/vnd.apple.mpegurl")
    public ResponseEntity<Resource> manifest(@PathVariable String id) throws Exception {
        transcode.ensureTranscoded(id);
        Path master = transcode.streamsDir().resolve(id).resolve("manifest.m3u8");
        if (!Files.exists(master)) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok()
            .header(HttpHeaders.CACHE_CONTROL, "no-cache")
            .header(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
            .body(new FileSystemResource(master));
    }

    @GetMapping("/{id}/{quality}/{file}")
    public ResponseEntity<Resource> file(
            @PathVariable String id,
            @PathVariable String quality,
            @PathVariable String file) {
        Path target = transcode.streamsDir().resolve(id).resolve(quality).resolve(file).normalize();
        Path root = transcode.streamsDir().resolve(id).normalize();
        if (!target.startsWith(root) || !Files.exists(target)) {
            return ResponseEntity.status(404)
                .body(new org.springframework.core.io.ByteArrayResource(
                    ("{\"id\":\"" + id + "\",\"quality\":\"" + quality + "\",\"segment\":\"" + file + "\",\"error\":\"not found\"}").getBytes()));
        }
        if (file.endsWith(".m3u8")) {
            return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("application/vnd.apple.mpegurl"))
                .header(HttpHeaders.CACHE_CONTROL, "no-cache")
                .header(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
                .body(new FileSystemResource(target));
        }
        // video/mp2t is the IANA media type for MPEG-2 Transport Stream segments.
        return ResponseEntity.ok()
            .contentType(MediaType.parseMediaType("video/mp2t"))
            .cacheControl(CacheControl.maxAge(Duration.ofDays(365)).cachePublic())
            .header(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
            .body(new FileSystemResource(target));
    }
}
